### PR TITLE
Structure_Engine: Add engine methods for retaining wall classes

### PR DESCRIPTION
**Pull request by KalleEdstroem**
### NOTE: Depends on 

https://github.com/BHoM/BHoM/pull/1621
   
### Issues addressed by this PR

Closes [BHoM_Structure_oM](https://github.com/BHoM/BHoM/issues/1619#issue-2388764920)

### Test files

See. 
https://github.com/BHoM/BHoM/pull/1621

### Changelog

Added required extension methods for implemented interfaces for new retaining wall classes.
- Added `Geometry()` method for `Stem` and `RetainingWall`
- Added `IsNull()` method for `Stem` and `RetainingWall`
- Added `SolidVolume()` method for `Stem` and `RetainingWall`
- Added `MaterialComposition()` method for `Stem` and `RetainingWall`
- Added `SetOutlineElements(List<IElement1D> outlineElements1D)` method for `Stem`
- Added `OutlineElemets1D()` method for `Stem`
- Added `Geometry3D()` method for `Stem` and `RetainingWall`
- Added New method for splitting a line into segments. `SplitIntoSegments`

### Additional comments

`HasMargeablePropertiesWith(IElement2D)` has not been added for any of the objects in reference to:
https://github.com/BHoM/BHoM_Engine/pull/3138#discussion_r1312778359

### CODE REVIEW

**Review by peterjamesnugent — CHANGES_REQUESTED** [review]
I have reviewed the code and done some basic checks in the script provided.

File: Structure_Engine/Query/IsNull.cs [hunk 1]
```diff
@@ -638,6 +638,72 @@ public static bool IsNull(this PileFoundation pileFoundation, [CallerMemberName]
...
+        [Description("Checks if a Footing is null and outputs relevant error message.")]
+        [Input("footing", "The Footing to test for null.")]
+        [Input("methodName", "The name of the method to reference in the error message.")]
+        [Input("msg", "Optional message to be returned in addition to the generated error message.")]
+        [Output("isNull", "True if the Footing is null.")]
+        public static bool IsNull(this Footing footing, [CallerMemberName] string methodName = "Method", string msg = "")
+        {
+            if (footing == null)
+            {
+                ErrorMessage(methodName, "PadFoundation", msg);
```
[the remark this answers is not in the document]
Comment (KalleEdstroem) [reply Structure_Engine/Query/IsNull.cs]: Mistake. updated now.

File: Structure_Engine/Query/IsNull.cs [hunk 2]
```diff
@@ -638,6 +638,72 @@ public static bool IsNull(this PileFoundation pileFoundation, [CallerMemberName]
...
+        [Description("Checks if a RetainingWall is null and outputs relevant error message.")]
+        [Input("retainingWall", "The RetainingWall to test for null.")]
+        [Input("methodName", "The name of the method to reference in the error message.")]
+        [Input("msg", "Optional message to be returned in addition to the generated error message.")]
+        [Output("isNull", "True if the PileFoundation is null.")]
+        public static bool IsNull(this RetainingWall retainingWall, [CallerMemberName] string methodName = "Method", string msg = "")
+        {
+            if (retainingWall == null)
+            {
+                ErrorMessage(methodName, "PadFoundation", msg);
```
[the remark this answers is not in the document]
Comment (KalleEdstroem) [reply Structure_Engine/Query/IsNull.cs:674]: Mistake. updated now.

File: Structure_Engine/Query/MaterialComposition.cs [hunk 3]
```diff
@@ -36,6 +36,7 @@
 using BH.oM.Dimensional;
 using BH.oM.Structure.Fragments;
 using BH.Engine.Base;
+using Microsoft.SqlServer.Server;
```
[the remark this answers is not in the document]
Comment (KalleEdstroem) [reply Structure_Engine/Query/MaterialComposition.cs:39]: No. Mistake, erased it now

File: Structure_Engine/Query/IsNull.cs [hunk 4]
```diff
@@ -638,6 +638,72 @@ public static bool IsNull(this PileFoundation pileFoundation, [CallerMemberName]
...
+        [Description("Checks if a Stem is null and outputs relevant error message.")]
+        [Input("stem", "The Stem to test for null.")]
+        [Input("methodName", "The name of the method to reference in the error message.")]
+        [Input("msg", "Optional message to be returned in addition to the generated error message.")]
+        [Output("isNull", "True if the PileFoundation is null.")]
+        public static bool IsNull(this Stem stem, [CallerMemberName] string methodName = "Method", string msg = "")
+        {
+            if (stem == null)
+            {
+                ErrorMessage(methodName, "PadFoundation", msg);
```
[the remark this answers is not in the document]
Comment (KalleEdstroem) [reply Structure_Engine/Query/IsNull.cs:651]: Mistake. Updated to `"Stem"`

File: Structure_Engine/Query/MaterialComposition.cs [hunk 5]
```diff
@@ -402,6 +403,84 @@ public static MaterialComposition MaterialComposition(this PileFoundation pileFo
...
+            {
+                ReinforcementDensity reinfDensity = retainingWall.FindFragment<ReinforcementDensity>();
+
+                List<IElementM> reinforcedElements = new List<IElementM>
+                {
+                    (IElementM)retainingWall.Stem.AddFragment(reinfDensity),
+                    (IElementM)retainingWall.Footing.AddFragment(reinfDensity)
+                };
+
+                return Matter.Compute.AggregateMaterialComposition(reinforcedElements);
```
[the remark this answers is not in the document]
Comment (KalleEdstroem) [reply Structure_Engine/Query/MaterialComposition.cs:463]: After the change to PadFoundation definition of footing the outline is refereed to as TopOutline. If a check is done to make sure that the bottom of the stem is on the same height as footing that should cover it right.

File: Structure_Engine/Create/Elements/RetainingWall.cs [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+        [Description("Creates a structural RetainingWall from a stem, a footing and properties.")]
+        [Input("stem", "The stem of the retaining wall.")]
+        [Input("footing", "The footing of the retaining wall.")]
+        [Input("retainedHeight", "The retained height of soil measured from the bottom of the wall Footing.")]
+        [Input("coverDepth", "The distance from top of Footing to finished floor level on the exposed face.")]
+        [Input("groundWaterDepth", "The distance from the base of the Footing to ground water level.")]
+        [Input("retentionAngle", "A property of the material being retained measured from the horizontal plane.")]
+        [Output("retainingWall", "The created RetainingWall containing the stem and footing.")]
+        public static RetainingWall RetainingWall(Stem stem, Footing footing, double retainedHeight = 0.0, double coverDepth = 0.0, double groundWaterDepth = 0.0, double retentionAngle = 0.0)
```
[the remark this answers is not in the document]
Comment (KalleEdstroem) [reply Structure_Engine/Create/Elements/RetainingWall.cs]: I understand. But There are default values in the oM when the Object is initialised... Do we want the to remove them as well and not let the object initialise without some/all properties? Just trying to understand it fully.

[the remark this answers is not in the document]
Comment (KalleEdstroem) [reply Structure_Engine/Create/Elements/RetainingWall.cs]: Discussion had and agreed to remove most of the default values.

File: Structure_Engine/Create/Elements/RetainingWall.cs [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+        [Input("footing", "The footing of the retaining wall.")]
+        [Output("retainingWall", "The created RetainingWall containing the stem and footing.")]
+        public static RetainingWall RetainingWall(Stem stem, Footing footing)
+        {
+            return new RetainingWall()
+            {
+                Stem = stem,
+                Footing = footing
+            };
+        }
```
[the remark this answers is not in the document]
Comment (KalleEdstroem) [reply Structure_Engine/Create/Elements/RetainingWall.cs]: Yes not really... Updated to a create method based on a line.

File: Structure_Engine/Query/Geometry3D.cs [hunk 8]
```diff
@@ -158,6 +159,108 @@ public static IGeometry Geometry3D(this PileFoundation pileFoundation)
 
             return compositeGeometry;
         }
+
+        /***************************************************/
+
+        [Description("Gets a CompositeGeometry made of boundary surfaces of the Footing based on its Outline and thickness.")]
```
[the remark this answers is not in the document]
Comment (KalleEdstroem) [reply Structure_Engine/Query/Geometry3D.cs]: Outdated.

File: Structure_Engine/Query/Geometry3D.cs [hunk 9]
```diff
@@ -158,6 +159,108 @@ public static IGeometry Geometry3D(this PileFoundation pileFoundation)
...
+
+            PlanarSurface botSrf = Engine.Geometry.Create.PlanarSurface(footing.Outline);
+
+
+            double thickness = footing.Property.ITotalThickness();
+            Vector extrudeVect = new Vector() { Z = thickness };
+
+            PlanarSurface topSrf = botSrf.ITranslate(extrudeVect) as PlanarSurface;
+
+            Extrusion externalEdgesExtrusion = Engine.Geometry.Create.Extrusion(footing.Outline, extrudeVect);
```
[the remark this answers is not in the document]
Comment (KalleEdstroem) [reply Structure_Engine/Query/Geometry3D.cs]: Outdated after update to PadFoundation.

File: Structure_Engine/Query/Geometry3D.cs [hunk 10]
```diff
@@ -158,6 +159,108 @@ public static IGeometry Geometry3D(this PileFoundation pileFoundation)
...
+            PlanarSurface frontSrf = centralPlanarSrf.ITranslate(stem.Normal.Normalise() * stem.ThicknessBottom / 2) as PlanarSurface;
+
+            //Figure out angle to rotate frontface by. Do some trig magic here. 
+            double a = stem.ThicknessBottom - stem.ThicknessTop;
+            double b = 5; // This is a palceholder value and will be changed.
+            double rad = Math.Atan(a / b);
+
+            //Lowest point. To rotrate around.
+            Point rotatePF = frontSrf.ControlPoints().OrderBy(p => p.Z).First();
+            Point rotatePB = backSrf.ControlPoints().OrderBy(p => p.Z).First();
```
[the remark this answers is not in the document]
Comment (KalleEdstroem) [reply Structure_Engine/Query/Geometry3D.cs:187]: I can't seem to find a method for rotating about a line. I will look a bit more and if not lets resolve it on monday.

**Review by KalleEdstroem — COMMENTED** [review]
This is a useful function but I can see the argument that it is too convoluted for the geometry engine and that methods here should be more straight forward.

The inputs `graceLength` and `fractionalDigits` could easily be removed and the method would function well. 

But, I still the merit of having those option and left it in for now.

File: Geometry_Engine/Modify/SplitIntoSegments.cs [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+        }
+
+        [Description("Splits a line in equal segments with gaps between so that a segment does not exceed the split interval.")]
+        [Input("line", "Line to be split in to segements if it exceeds the split interval.")]
+        [Input("splitInterval", "Length that the segments should not exceed.")]
+        [Input("gapLength", "Length of the gap between each segment.")]
+        [Input("graceLength", "If the total length of the wall is lower than this number it will not be segmented. The whole line is returned.")]
+        [Input("fractionalDigits", "The number of fractional digits to round the segment length to. An integer between 0 and 15 inclusive. Zero will round to nearest 5/0.")]
+        [Output("segments", "Segments of input line, not exceeding the split interval.")]
+        public static List<Line> SplitInToSegments(Line line, double splitInterval, double gapLength = 0.0, double graceLength = 0.0, int fractionalDigits = -1)
```
Comment (KalleEdstroem) [inline Geometry_Engine/Modify/SplitIntoSegments.cs:70]: This is a useful function but I can see the argument that it is too convoluted for the geometry engine and that methods here should be more straight forward.

The inputs `graceLength` and `fractionalDigits` could easily be removed and the method would function well.

But, I still the merit of having those option and left it in for now.

File: Structure_Engine/Query/Geometry3D.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (peterjamesnugent) [reply Structure_Engine/Query/Geometry3D.cs]: I think you've resolved this, I can see you're rotating about a Vector?

File: Structure_Engine/Query/MaterialComposition.cs [hunk 12]
```diff
@@ -402,6 +402,69 @@ public static MaterialComposition MaterialComposition(this PileFoundation pileFo
...
+
+
+            List<IElementM> elements = new List<IElementM>
+            {
+                retainingWall.Stem,
+                retainingWall.Footing
+            };
+
+            //Case for when ReinforcementDensity is attached to both hihgher and lower level objects.
+            if ((!retainingWall.Stem.FindFragment<ReinforcementDensity>().IsNull() || !retainingWall.Footing.FindFragment<ReinforcementDensity>().IsNull()) && !retainingWall.FindFragment<ReinforcementDensity>().IsNull())
```
Comment (peterjamesnugent) [inline Structure_Engine/Query/MaterialComposition.cs:439]:
```suggestion
            if ((retainingWall.Stem.FindFragment<ReinforcementDensity>() != null || retainingWall.Footing.FindFragment<ReinforcementDensity>() != null) && retainingWall.FindFragment<ReinforcementDensity>() != null)
```
When you call `IsNull` this is why you are getting the error message despite the method working because it raises an auto warning. Here I would just check `!= null`, it's a bit easier to read that `!(object == null)`.

Comment (KalleEdstroem) [reply Structure_Engine/Query/MaterialComposition.cs:439]: Cool. Thank you!

File: Structure_Engine/Query/Geometry3D.cs [hunk 13]
```diff
@@ -158,6 +159,108 @@ public static IGeometry Geometry3D(this PileFoundation pileFoundation)
...
+
+            PlanarSurface centralPlanarSrf = Engine.Geometry.Create.PlanarSurface(stem.Outline);
+
+            PlanarSurface backSrf = centralPlanarSrf.ITranslate(stem.Normal.Normalise() * -stem.ThicknessBottom / 2) as PlanarSurface;
+            PlanarSurface frontSrf = centralPlanarSrf.ITranslate(stem.Normal.Normalise() * stem.ThicknessBottom / 2) as PlanarSurface;
+
+            //Figure out angle to rotate frontface by. Do some trig magic here. 
+            double a = stem.ThicknessBottom - stem.ThicknessTop;
+            double b = 5; // This is a palceholder value and will be changed.
+            double rad = Math.Atan(a / b);
```
[the remark this answers is not in the document]
Comment (KalleEdstroem) [reply Structure_Engine/Query/Geometry3D.cs]: N/A anymore. Geometry3D() simplified.

File: Structure_Engine/Query/Geometry3D.cs [hunk 14]
```diff
@@ -158,6 +159,108 @@ public static IGeometry Geometry3D(this PileFoundation pileFoundation)
...
+
+            frontSrf = frontSrf.IRotate(rotatePF, stem.Normal.CrossProduct(Vector.ZAxis), rad) as PlanarSurface;
+            backSrf = backSrf.IRotate(rotatePB, stem.Normal.CrossProduct(Vector.ZAxis), -rad) as PlanarSurface;
+
+            //Extend the surface to be at correct height with the walls. Need to be extended dependant on the rotation angle.
+            //is there a method to extend it? (Have one in my back pocket though...)
+
+            //Close out the shape. Would be great with a method to close it out...
+
+            // Creates an extrusion closing the shape. Needs to be dut by the other surface.
```
[the remark this answers is not in the document]
Comment (KalleEdstroem) [reply Structure_Engine/Query/Geometry3D.cs]: N/A. Geometry3D() simplified.

File: Physical_Engine/Query/VolumetricMaterialTakeoff.cs [hunk 15]
```diff
@@ -30,7 +30,6 @@
 using System;
 using System.Collections.Generic;
 using System.ComponentModel;
-using System.Linq;
```
Comment (peterjamesnugent) [inline Physical_Engine/Query/VolumetricMaterialTakeoff.cs:33]: Can you revert this change, we shouldn't be making any changes to the Physical_Engine in a Structure_Engine PR.

File: Structure_Engine/Create/Elements/Stem.cs [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+
+namespace BH.Engine.Structure
+{
+    public static partial class Create
+    {
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        [Description("Creates a tapered Stem.")]
```
Comment (peterjamesnugent) [inline Structure_Engine/Create/Elements/Stem.cs:43]:
```suggestion
        [Description("Creates a tapered Stem to be used in the RetainingWall object.")]
```

**Comment by KalleEdstroem** [comment]
@BHoMBot check compliance

File: Structure_Engine/Modify/Flip.cs [hunk 17]
```diff
@@ -69,6 +69,26 @@ public static Edge Flip(this Edge edge)
...
+        [Output("stem", "The Stem with flipped Normal")]
+        public static Stem Flip(this Stem stem)
+        {
+            if (stem.IsNull())
+                return null;
+            if (stem.Normal.IsNull())
+            {
+                Base.Compute.RecordError("The Normal of the Stem is null and could not be flipped.");
+                return null;
+            }
```
Comment (peterjamesnugent) [inline Structure_Engine/Modify/Flip.cs:84]:
```suggestion
            if (stem.Normal.IsNull("The Normal of the Stem is null and could not be flipped.","Flip"))
                return null;
```
The `IsNull` has an overload where you can provide a message, so this can be simplified.

File: Structure_Engine/Query/Geometry.cs [hunk 18]
```diff
@@ -121,6 +121,31 @@ public static IGeometry Geometry(this PileFoundation pileFoundation)
             return Engine.Geometry.Create.CompositeGeometry(geometry);
         }
 
+        /***************************************************/
+
+        [Description("Gets the geometry of a Stem. Method required for automatic display in UI packages.")]
+        [Input("stem", "Stem to get the surface geometry from.")]
```
Comment (peterjamesnugent) [inline Structure_Engine/Query/Geometry.cs:127]:
```suggestion
        [Input("stem", "Stem to get the geometry from.")]
```

File: Structure_Engine/Query/Geometry3D.cs [hunk 19]
```diff
@@ -158,6 +158,55 @@ public static IGeometry Geometry3D(this PileFoundation pileFoundation)
...
+        [Input("stem", "The input Stem to get the Geometry3D out of.")]
+        [Output("3d", "Three-dimensional geometry of the Stem.")]
+        public static IGeometry Geometry3D(this Stem stem)
+        {
+            if (stem.IsNull())
+                return null;
+
+            if (stem.ThicknessBottom != stem.ThicknessTop)
+            {
+                Base.Compute.RecordError("The Geometry3D() method is not implemented for a tapered Stem. Null is returned.");
```
Comment (peterjamesnugent) [inline Structure_Engine/Query/Geometry3D.cs:174]:
```suggestion
                Base.Compute.RecordError("The Geometry3D() method is not implemented for a tapered Stem.");
```

File: Structure_Engine/Query/IsNull.cs [hunk 20]
```diff
@@ -638,6 +638,49 @@ public static bool IsNull(this PileFoundation pileFoundation, [CallerMemberName]
 
             return false;
         }
+        
+        [Description("Checks if a Stem is null and outputs relevant error message.")]
+        [Input("stem", "The Stem to test for null.")]
+        [Input("methodName", "The name of the method to reference in the error message.")]
+        [Input("msg", "Optional message to be returned in addition to the generated error message.")]
+        [Output("isNull", "True if the PileFoundation is null.")]
```
Comment (peterjamesnugent) [inline Structure_Engine/Query/IsNull.cs:646]:
```suggestion
        [Output("isNull", "True if the Stem is null.")]
```

File: Structure_Engine/Query/IsNull.cs [hunk 21]
```diff
@@ -638,6 +638,49 @@ public static bool IsNull(this PileFoundation pileFoundation, [CallerMemberName]
...
+                return true;
+
+            return false;
+        }
+
+        [Description("Checks if a RetainingWall is null and outputs relevant error message.")]
+        [Input("retainingWall", "The RetainingWall to test for null.")]
+        [Input("methodName", "The name of the method to reference in the error message.")]
+        [Input("msg", "Optional message to be returned in addition to the generated error message.")]
+        [Output("isNull", "True if the PileFoundation is null.")]
```
Comment (peterjamesnugent) [inline Structure_Engine/Query/IsNull.cs:669]:
```suggestion
        [Output("isNull", "True if the RetainingWall is null.")]
```

**Review by peterjamesnugent — CHANGES_REQUESTED** [review]
I need to verify the different calculation methods, but here are some comments for now.

File: Structure_Engine/Modify/Flip.cs (hunk 17 above)
Comment (KalleEdstroem) [reply Structure_Engine/Modify/Flip.cs:84]: Perfect , thanks!

File: Structure_Engine/Query/IsNull.cs (hunk 20 above)
Comment (KalleEdstroem) [reply Structure_Engine/Query/IsNull.cs]: thanks

File: Structure_Engine/Query/IsNull.cs (hunk 21 above)
Comment (KalleEdstroem) [reply Structure_Engine/Query/IsNull.cs:669]: thanks

File: Physical_Engine/Query/VolumetricMaterialTakeoff.cs (hunk 15 above)
Comment (KalleEdstroem) [reply Physical_Engine/Query/VolumetricMaterialTakeoff.cs]: Thanks for picking that up. It is an accidental change. Updated per: [87d429a](https://github.com/BHoM/BHoM_Engine/pull/3367/commits/87d429a297116cc8962e0dfe886be8b3bda7e40c)

File: Structure_Engine/Query/SolidVolume.cs [hunk 22]
```diff
@@ -106,6 +104,34 @@ public static double SolidVolume(this PileFoundation pileFoundation)
...
+        /***************************************************/
+
+        [Description("Returns a RetainingWall's solid volume based on the Stem and Footing volumes.")]
+        [Input("retainingWall", "The RetainingWall to get the volume from.")]
+        [Output("volume", "The RetainingWall solid material volume.", typeof(Volume))]
+        public static double SolidVolume(this RetainingWall retainingWall)
+        {
+            if (retainingWall.Footing.TopOutline.ControlPoints().OrderBy(p => p.Z).First().Z < retainingWall.Stem.Outline.ControlPoints().OrderBy(p => p.Z).First().Z)
+            {
+                Base.Compute.RecordError("The footings highest control point is above the lowest control point of the stem. The two objects should not go into eachother. A NaN is returned.");
```
Comment (peterjamesnugent) [inline Structure_Engine/Query/SolidVolume.cs:126]:
```suggestion
                Base.Compute.RecordError("The footings highest control point is above the lowest control point of the stem. The two objects should not go into eachother. ");
```

File: Structure_Engine/Query/SolidVolume.cs [hunk 23]
```diff
@@ -106,6 +104,34 @@ public static double SolidVolume(this PileFoundation pileFoundation)
...
+        }
+
+        /***************************************************/
+
+        [Description("Returns a RetainingWall's solid volume based on the Stem and Footing volumes.")]
+        [Input("retainingWall", "The RetainingWall to get the volume from.")]
+        [Output("volume", "The RetainingWall solid material volume.", typeof(Volume))]
+        public static double SolidVolume(this RetainingWall retainingWall)
+        {
+            if (retainingWall.Footing.TopOutline.ControlPoints().OrderBy(p => p.Z).First().Z < retainingWall.Stem.Outline.ControlPoints().OrderBy(p => p.Z).First().Z)
```
Comment (peterjamesnugent) [inline Structure_Engine/Query/SolidVolume.cs:124]:
```suggestion
        [Description("Returns a RetainingWall's solid volume by summing the volume of the Stem and Footing volumes.")]
        [Input("retainingWall", "The RetainingWall to get the volume from.")]
        [Output("volume", "The RetainingWall solid material volume.", typeof(Volume))]
        public static double SolidVolume(this RetainingWall retainingWall)
        {
            if (retainingWall.Footing.TopOutline.ControlPoints().OrderBy(p => p.Z).First().Z < retainingWall.Stem.Outline.ControlPoints().OrderBy(p => p.Z).First().Z)
```
We need to add tolerances for this if we are going to check this. I am of the opinion that we keep the method simple and adjust the descriptions appropriately. 

We can also have an `IsValid()` method in the future to check if the `RetainingWall` overlaps etc.

File: Structure_Engine/Query/SolidVolume.cs [hunk 24]
```diff
@@ -106,6 +104,34 @@ public static double SolidVolume(this PileFoundation pileFoundation)
...
+            {
+                Base.Compute.RecordError("The footings highest control point is above the lowest control point of the stem. The two objects should not go into eachother. A NaN is returned.");
+                return double.NaN;
```
Comment (peterjamesnugent) [inline Structure_Engine/Query/SolidVolume.cs:127]:
```suggestion
                return 0;
```

File: Structure_Engine/Create/Elements/RetainingWall.cs [hunk 25]
```diff
@@ -0,0 +1,130 @@
...
+            {
+                Base.Compute.RecordError("The stem is not connected to the footing. Make sure the stem bottom is on the footing outline.");
+                return null;
+            }
+
+            if (stem.IsNull())
+                return null;
+
+            if (footing.IsNull())
+                return null;
```
Comment (peterjamesnugent) [inline Structure_Engine/Create/Elements/RetainingWall.cs:73]: This should be the first check you do, L55 will return a `NullException` if the `Stem` is `null`.

File: Structure_Engine/Create/Elements/RetainingWall.cs [hunk 26]
```diff
@@ -0,0 +1,130 @@
...
+                Point end = curve.IEndPoint();
+                curves.Add(curve);
+            }
+            ICurve bottomCurve = curves.OrderBy(p => p.IStartPoint().Z + p.IEndPoint().Z).First();
+
+            if (!footing.TopOutline.IsContaining(bottomCurve))
+            {
+                Base.Compute.RecordError("The stem is not connected to the footing. Make sure the stem bottom is on the footing outline.");
+                return null;
+            }
```
Comment (peterjamesnugent) [inline Structure_Engine/Create/Elements/RetainingWall.cs:67]: As with the other method, I would put this in an `IsValid` method.

File: Structure_Engine/Query/MaterialComposition.cs [hunk 27]
```diff
@@ -402,6 +402,69 @@ public static MaterialComposition MaterialComposition(this PileFoundation pileFo
             return Matter.Compute.AggregateMaterialComposition(elements);
         }
 
+        /***************************************************/
+
+        [Description("Returns a Stem's homogeneous MaterialComposition.")]
+        [Input("stem", "The Stem to get material from.")]
```
Comment (peterjamesnugent) [inline Structure_Engine/Query/MaterialComposition.cs:408]:
```suggestion
        [Input("stem", "The Stem to query.")]
```

File: Structure_Engine/Query/MaterialComposition.cs [hunk 28]
```diff
@@ -402,6 +402,69 @@ public static MaterialComposition MaterialComposition(this PileFoundation pileFo
...
+
+            ReinforcementDensity reinfDensity = stem.FindFragment<ReinforcementDensity>();
+
+            return MaterialComposition(stem.Material, reinfDensity);
+        }
+
+        /***************************************************/
+
+        [Description("Returns a RetainingWall's homogeneous MaterialComposition based on the Stem and Footing.")]
+        [Input("retainingWall", "The RetainingWall to get material from.")]
```
Comment (peterjamesnugent) [inline Structure_Engine/Query/MaterialComposition.cs:423]:
```suggestion
        [Input("retainingWall", "The RetainingWall to query.")]
```

File: Structure_Engine/Query/MaterialComposition.cs [hunk 29]
```diff
@@ -402,6 +402,69 @@ public static MaterialComposition MaterialComposition(this PileFoundation pileFo
             return Matter.Compute.AggregateMaterialComposition(elements);
         }
 
+        /***************************************************/
+
+        [Description("Returns a Stem's homogeneous MaterialComposition.")]
+        [Input("stem", "The Stem to get material from.")]
+        [Output("materialComposition", "The kind of matter the Stem is composed of.")]
```
Comment (peterjamesnugent) [inline Structure_Engine/Query/MaterialComposition.cs:409]:
```suggestion
        [Output("materialComposition", "The MaterialComposition of the Stem")]
```

File: Structure_Engine/Query/MaterialComposition.cs [hunk 30]
```diff
@@ -402,6 +402,69 @@ public static MaterialComposition MaterialComposition(this PileFoundation pileFo
...
+            ReinforcementDensity reinfDensity = stem.FindFragment<ReinforcementDensity>();
+
+            return MaterialComposition(stem.Material, reinfDensity);
+        }
+
+        /***************************************************/
+
+        [Description("Returns a RetainingWall's homogeneous MaterialComposition based on the Stem and Footing.")]
+        [Input("retainingWall", "The RetainingWall to get material from.")]
+        [Output("materialComposition", "The kind of matter the RetainingWall is composed of.")]
```
Comment (peterjamesnugent) [inline Structure_Engine/Query/MaterialComposition.cs:424]:
```suggestion
        [Output("materialComposition", "The MaterialComposition of the RetainingWall")]
```

**Review by peterjamesnugent — CHANGES_REQUESTED** [review]
Few more comments.

File: Structure_Engine/Create/Elements/RetainingWall.cs (hunk 25 above)
Comment (KalleEdstroem) [reply Structure_Engine/Create/Elements/RetainingWall.cs:73]: Thanks, updated.

File: Structure_Engine/Query/SolidVolume.cs (hunk 23 above)
Comment (KalleEdstroem) [reply Structure_Engine/Query/SolidVolume.cs]: Added an IsValid method and implemented it here.

File: Structure_Engine/Create/Elements/RetainingWall.cs (hunk 26 above)
Comment (KalleEdstroem) [reply Structure_Engine/Create/Elements/RetainingWall.cs]: Added IsValid method taking a stem and a footing and implementing it here.

**Comment by peterjamesnugent** [comment]
@BHoMBot check compliance

File: Structure_Engine/Create/Elements/RetainingWall.cs [hunk 31]
```diff
@@ -0,0 +1,116 @@
...
+
+            if (Query.IsValid(stem, footing))
+                return new RetainingWall() { Stem = stem, Footing = footing, RetainedHeight = retainedHeight, CoverDepth = coverDepth, RetentionAngle = retentionAngle, GroundWaterDepth = groundWaterDepth };
+            else
+                return null;
+        }
+
+        /***************************************************/
+
+        //TODO Add input descriptions here
```
Comment (peterjamesnugent) [inline Structure_Engine/Create/Elements/RetainingWall.cs:66]: Complete this please, can you be clear that the line needs to exist in a single plane.

File: Structure_Engine/Create/Elements/RetainingWall.cs [hunk 32]
```diff
@@ -0,0 +1,116 @@
...
+            ConstantThickness footingThickness,
+            double coverDepth,
+            double retentionAngle
+            )
+        {
+            if (line.Start == null && line.End == null) { return null; }
+
+            if (!line.IsInPlane(Geometry.Create.Plane(line.Start, Vector.ZAxis)))
+            {
+                Base.Compute.RecordError("Provided line is not parallell to the XY plane. Please provide a line parallell to the XY plane.");
```
Comment (peterjamesnugent) [inline Structure_Engine/Create/Elements/RetainingWall.cs:85]:
```suggestion
                Base.Compute.RecordError("Provided line is not within a single Plane. Please provide a line that exists within in a single plane.");
```

File: Structure_Engine/Create/Elements/RetainingWall.cs [hunk 33]
```diff
@@ -0,0 +1,116 @@
...
+                return null; 
+            }
+
+            PolyCurve stemOutline = new PolyCurve();
+            PolyCurve footingOutline = new PolyCurve();
+
+            //Create the footing outline. 
+            Line toeLine = line.DeepClone();
+            Line heelLine = line.DeepClone();
```
Comment (peterjamesnugent) [inline Structure_Engine/Create/Elements/RetainingWall.cs:94]:
```suggestion
            Line toeLine = line.ShallowClone();
            Line heelLine = line.ShallowClone();
```

File: Structure_Engine/Query/MaterialComposition.cs [hunk 34]
```diff
@@ -402,6 +402,69 @@ public static MaterialComposition MaterialComposition(this PileFoundation pileFo
...
+            List<IElementM> elements = new List<IElementM>
+            {
+                retainingWall.Stem,
+                retainingWall.Footing
+            };
+
+            //Case for when ReinforcementDensity is attached to both higher and lower level objects.
+            if ((retainingWall.Stem.FindFragment<ReinforcementDensity>() != null || retainingWall.Footing.FindFragment<ReinforcementDensity>() != null) && retainingWall.FindFragment<ReinforcementDensity>() != null)
+            {
+                Base.Compute.RecordWarning("A reinforcement density fragment is found on both the retaining wall object and on at least one of its defining objects. The reinforcementdensity of the lower level parts has been used.");
```
Comment (peterjamesnugent) [inline Structure_Engine/Query/MaterialComposition.cs:441]:
```suggestion
                Base.Compute.RecordWarning("A ReinforcementDensity Fragment is found on both the RetainingWall and on at least one of its defining objects. The ReinforcementDensity of the defining objects has been used.");
```

**Review by peterjamesnugent — CHANGES_REQUESTED** [review]

File: Structure_Engine/Create/Elements/RetainingWall.cs (hunk 32 above)
Comment (KalleEdstroem) [reply Structure_Engine/Create/Elements/RetainingWall.cs:93]: Stated in descriptions that it should be parallell to XY plane

**Comment by peterjamesnugent** [comment]
@BHoMBot check compliance

**Comment by peterjamesnugent** [comment]
@BHoMBot check compliance

**Comment by peterjamesnugent** [comment]
@BHoMBot check required

**Comment by peterjamesnugent** [comment]
@BHoMBot check unit-tests
@BHoMBot check dataset-compliance
@BHoMBot check copyright-compliance

**Comment by peterjamesnugent** [comment]
@BHoMBot check unit-tests

**Comment by peterjamesnugent** [comment]
@BHoMBot check unit-tests

**Comment by peterjamesnugent** [comment]
@BHoMBot check unit-tests
@BHoMBot check versioning

**Comment by peterjamesnugent** [comment]
@BHoMBot check unit-tests

**Comment by peterjamesnugent** [comment]
The check unit-tests appears to be failing despite rebasing with the changes to the Search_Engine. These changes should result in passing unit-test check as evidenced by a PR raised post-merge of the Search_Engine: https://github.com/BHoM/BHoM_Engine/pull/3396/checks?check_run_id=29376096738

I am going to close this PR, rename the branch on the remote to [REMOVED]Structure_oM-#1619-AddRetainingWallClasses and raise a new branch from \develop called Structure_oM-#1619-AddRetainingWallClasses (so the installer/core checks can work elsewhere).
